File: mobile/ios/SpeakeasyCrypto/Vault/VaultTestRunner.swift
import Foundation

// Test harness for Vault operations (Import -> Encrypt -> Store -> Load -> Decrypt)

class VaultTestRunner {
    
    static func run() {
        print("🚀 Starting Vault Test...")
        
        let store = KeychainStore()
        let keyManager = VaultKeyManager(store: store)
        let fileStore = VaultFileStore()
        let indexDb = VaultIndexDb()
        
        do {
            // 0. Ensure identity exists (for DMK derivation)
            if store.getIdentityKeyPair() == nil {
                print("⚠️ No identity key found. Run RoundtripRunner first to generate identity.")
                return
            }
            
            // 1. Create test plaintext
            let testPlaintext = "Hello Speakeasy Vault! 🔐".data(using: .utf8)!
            print("✅ Plaintext: \(String(data: testPlaintext, encoding: .utf8)!)")
            
            // 2. Generate FileKey and encrypt
            let fileKey = keyManager.generateFileKey()
            let (ciphertext, nonce) = try AttachmentCrypto.encrypt(plaintext: testPlaintext, key: fileKey)
            print("✅ Encrypted: \(ciphertext.count) bytes")
            
            // 3. Wrap FileKey
            let wrappedKey = try keyManager.wrapFileKey(fileKey)
            print("✅ FileKey wrapped: \(wrappedKey.count) bytes")
            
            // 4. Save to disk
            let itemId = UUID()
            let _ = try fileStore.save(ciphertext: ciphertext, id: itemId)
            print("✅ Saved to disk: \(itemId)")
            
            // 5. Save to index
            let vaultItem = VaultItem(
                id: itemId,
                type: .note,
                createdAt: Date(),
                updatedAt: Date(),
                wrappedFileKeyB64: wrappedKey.base64EncodedString(),
                nonceB64: nonce.base64EncodedString(),
                ciphertextPath: "\(itemId).enc",
                attachmentId: nil,
                mimeType: "text/plain",
                ciphertextSizeBytes: Int64(ciphertext.count),
                displayNameEnc: nil
            )
            indexDb.insert(vaultItem)
            print("✅ Saved to index")
            
            // 6. Simulate "reboot" - load from disk
            let loadedItem = indexDb.get(itemId)!
            let loadedCiphertext = try fileStore.load(id: itemId)
            print("✅ Loaded from disk: \(loadedCiphertext.count) bytes")
            
            // 7. Unwrap FileKey
            let unwrappedKey = try keyManager.unwrapFileKey(wrapped: Data(base64Encoded: loadedItem.wrappedFileKeyB64!)!)
            print("✅ FileKey unwrapped")
            
            // 8. Decrypt
            let loadedNonce = Data(base64Encoded: loadedItem.nonceB64)!
            let decrypted = try AttachmentCrypto.decrypt(ciphertext: loadedCiphertext, nonce: loadedNonce, key: unwrappedKey)
            let decryptedString = String(data: decrypted, encoding: .utf8)!
            print("✅ Decrypted: \(decryptedString)")
            
            // 9. Verify
            if decryptedString == "Hello Speakeasy Vault! 🔐" {
                print("🏆 VAULT TEST SUCCESS!")
            } else {
                print("❌ MISMATCH")
            }
            
            // Cleanup
            try fileStore.delete(id: itemId)
            indexDb.delete(itemId)
            print("🧹 Cleaned up test data")
            
        } catch {
            print("❌ ERROR: \(error)")
        }
    }
}

File: mobile/android/crypto/vault/VaultTestRunner.kt
package com.speakeasy.crypto.vault

import android.content.Context
import android.util.Log

// Test harness for Vault operations (Import -> Encrypt -> Store -> Load -> Decrypt)

class VaultTestRunner(private val context: Context) {
    
    private val TAG = "VaultTestRunner"
    
    fun run() {
        Log.d(TAG, "🚀 Starting Vault Test...")
        
        val store = com.speakeasy.crypto.KeystoreStore(context)
        val keyManager = VaultKeyManager(store)
        val fileStore = VaultFileStore(context)
        val indexDb = VaultIndexDb(context)
        
        try {
            // 0. Ensure identity exists (for DMK derivation)
            if (store.getIdentityKeyPair() == null) {
                Log.w(TAG, "⚠️ No identity key found. Generating test identity...")
                // Generate a fake identity for testing
                val fakeIdentity = ByteArray(32)
                java.security.SecureRandom().nextBytes(fakeIdentity)
                store.saveIdentity(fakeIdentity, 12345)
            }
            
            // 1. Create test plaintext
            val testPlaintext = "Hello Speakeasy Vault! 🔐".toByteArray(Charsets.UTF_8)
            Log.d(TAG, "✅ Plaintext: ${String(testPlaintext)}")
            
            // 2. Generate FileKey and encrypt
            val fileKey = keyManager.generateFileKey()
            val (ciphertext, nonce) = AttachmentCrypto.encrypt(testPlaintext, fileKey)
            Log.d(TAG, "✅ Encrypted: ${ciphertext.size} bytes")
            
            // 3. Wrap FileKey
            val wrappedKey = keyManager.wrapFileKey(fileKey)
            Log.d(TAG, "✅ FileKey wrapped: ${wrappedKey.size} bytes")
            
            // 4. Save to disk
            val itemId = java.util.UUID.randomUUID()
            fileStore.save(ciphertext, itemId)
            Log.d(TAG, "✅ Saved to disk: $itemId")
            
            // 5. Save to index
            val vaultItem = VaultItem(
                id = itemId,
                type = VaultItemType.NOTE,
                createdAt = java.util.Date(),
                updatedAt = java.util.Date(),
                wrappedFileKeyB64 = android.util.Base64.encodeToString(wrappedKey, android.util.Base64.NO_WRAP),
                nonceB64 = android.util.Base64.encodeToString(nonce, android.util.Base64.NO_WRAP),
                ciphertextPath = "${itemId}.enc",
                attachmentId = null,
                mimeType = "text/plain",
                ciphertextSizeBytes = ciphertext.size.toLong(),
                displayNameEnc = null
            )
            indexDb.insert(vaultItem)
            Log.d(TAG, "✅ Saved to index")
            
            // 6. Simulate "reboot" - load from disk
            val loadedItem = indexDb.get(itemId)!!
            val loadedCiphertext = fileStore.load(itemId)
            Log.d(TAG, "✅ Loaded from disk: ${loadedCiphertext.size} bytes")
            
            // 7. Unwrap FileKey
            val unwrappedKey = keyManager.unwrapFileKey(
                android.util.Base64.decode(loadedItem.wrappedFileKeyB64, android.util.Base64.NO_WRAP)
            )
            Log.d(TAG, "✅ FileKey unwrapped")
            
            // 8. Decrypt
            val loadedNonce = android.util.Base64.decode(loadedItem.nonceB64, android.util.Base64.NO_WRAP)
            val decrypted = AttachmentCrypto.decrypt(loadedCiphertext, loadedNonce, unwrappedKey)
            val decryptedString = String(decrypted, Charsets.UTF_8)
            Log.d(TAG, "✅ Decrypted: $decryptedString")
            
            // 9. Verify
            if (decryptedString == "Hello Speakeasy Vault! 🔐") {
                Log.d(TAG, "🏆 VAULT TEST SUCCESS!")
            } else {
                Log.e(TAG, "❌ MISMATCH")
            }
            
            // Cleanup
            fileStore.delete(itemId)
            indexDb.delete(itemId)
            Log.d(TAG, "🧹 Cleaned up test data")
            
        } catch (e: Exception) {
            Log.e(TAG, "❌ ERROR: ${e.message}", e)
        }
    }
}
